### PR TITLE
Reduce transient memory allocations in integrate/pw_lin.

### PR DESCRIPTION
**Pull request by thorstenhater**
Another break-out of #2005 

- Fix logic bug in determining diffusivity, thanks @halfflat 
- Implement look-back-of-one caching for region/locset thingification
- Untangle `embed_pwlin` integration, save a bit of allocation.

Overall saves ~10% when creating 2048 complex cells of depth 
10 in `busyring`.

*NOTE* `pw_zip_*` and `pw_over_cable` still allocate tons of transients
as they use `pw_elements<pw_elements<double>>` ie a vector of vectors.
Due to the way this is constructed, dereferencing iterators on these types
returns their contained _value_ not _const value&_, thus _copying_ the inner
vectors each time. Overall -- despite the frustration and the complexity -- the
remainder doesn't seem worth addressing. For now.

### CODE REVIEW

File: arbor/util/piecewise.hpp [hunk 1]
```diff
@@ -250,6 +253,11 @@ struct pw_elements {
         counter<pw_size_type>& inner() { return c_; }
         const counter<pw_size_type>& inner() const { return c_; }
 
+        double lower_bound() const { return pw_->extent(*c_).first; }
+        double upper_bound() const { return pw_->extent(*c_).second; }
+
+        const X& value() const { return pw_->cvalue(*c_); };
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/util/piecewise.hpp:259]: Note that `reference` is not, in fact, a reference:
```cxx
        using reference = pw_element<X>;

        reference operator[](difference_type j) const { return (*pw_)[j+*c_]; }
        reference operator*() const { return (*pw_)[*c_]; }
        pointer operator->() const { return pointer{(*pw_)[*c_]}; }
```
thus, `[]` and `*` will construct a full `pw_element<X>`. This is usually not offensive.
BUT, in some cases `X=vector<double>`, which is obviously an issue. I couldn't get a
working instance where `reference = pw_element<X>&` or what we'd actually want here
`reference = pw_element<X&>`.

[the remark this answers is not in the document]
Comment (boeschf) [reply arbor/util/piecewise.hpp:259]: yes, that's why I proposed to `using reference = pw_element_proxy<const X>` and specialize
```c++
template<typename X>
pw_element_proxy<const X> { ... };
```

File: arbor/util/piecewise.hpp [hunk 2]
```diff
@@ -240,16 +259,21 @@ struct pw_elements {
 
         using value_type = pw_element<X>;
         using pointer = const pointer_proxy<pw_element<X>>;
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/util/piecewise.hpp:261]: This doesn't compile.

**Comment by thorstenhater** [comment]
Split into 
#2389 
#2388

**Pull request closed without merging**